### ISSUE TITLE
Bug: @typescript-eslint/no-unnecessary-condition false positive on indexed access

### ISSUE THREAD

**Issue by atombrenner** [typescript-eslint/typescript-eslint#10901]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=C4TwDgpgBAksEFsoF4oG8BQUoCMCGATgFxQDOwBAlgHYDmGAvhhgPQtQETACuB1pUPNUEECeEFADulYAAsoALwgEA9lBUEoCDdAgAbRBGrBSGAGbdqAY2CUVw0hEJXZACgCUJOIgDaAXXQsDi5eYX9GZit7cmDSbj0TFDInAhcPZjYoajVlVU1KMyhuUhpaKBoAEwgADwgKwSsrCFJTAqhXTjiE0h8ABj93QOwo-hUDADo9FVoO5viTPoGI1nZcjSlZI0E9SjwSuig5aDNKAhiZRAwRmJOz4G8kVE75nv6MNtdb8gfBzGHosYQSbTT6nb7wBDuZbXYBQHxfYABVCOZxuKEfBG-KCZMx4PSOKBgFQlWwANwgABpkqiPFpxMEeHxBMJEGBQCIxCAgtdAcCZpioBEgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDToMmeCmFHopAXxAqgA&tokens=false

### Repro Code

```TypeScript
type Item = {
  bar: string
}

// returns an array with zero or more elements
function search(): Item[] {
  return []
}

const [first] = search()
if (first) { // false positive, search() may return an empty array
  console.log(first) 
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true,
    "strict": true
  }
}
```

### Expected Result

I expected that the check for an array member, that can be undefined, is not reported as an unnecessary condition. Removing the check would lead to runtime errors (and quite a lot of devs trust eslint and will remove the check).
 

### Actual Result

```
const [first] = search()
if (first) { // false positive, search() may return an empty array
  console.log(first) 
}
```

### Additional Info

IMO the root cause for this problem is  that typescript with default strict settings infers the type of `first` as `Foo` while it should be `Foo | undefined`. We can fix this with the [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess) compiler option. 
By default this option is set to false, even if `"strict": true` is configured. Enabling this option makes working with arrays cumbersome so it's rarely enabled in projects. 

This rule is part of the strict rule set, but it works only correct if `noUncheckedIndexedAccess` is enabled *or* when the index operator is part of the condition. So projects that want to improve their code base by using a more rigid ruleset, might introduce  new bugs when following linter warnings.

To mitigate the problem, this behaviour should be documented in the rule description or maybe the rule can be improved to check the tsconfig and change behaviour accordingly.
It is inconsistent to not warn when the index operator [] is part condition, but warn if  the member is accesses via destructuring or assigning to another variable.

**Comment by bradzacher** [comment]
We have some logic to look at direct index access and avoid erroring on the `if (arr[idx])` case. We can easily see that it's index access based on the syntax, and we can inspect the type of `arr` to see that it's also an array.

The issue is that once you no longer inline the array element access things get harder to trace overall.
The TS types here aren't helpful to checking for this case and so we need to rely on other techniques to try to catch this.
We could probably do more here but we generally avoid relying on scope analysis to do deeper AST analysis because things get really complicated really fast.

We haven't delved into this more because there are two easy enough userland solutions here.
Either enable `noUncheckedIndexedAccess` or to use the `arr.at(idx)` method instead as it encodes the `undefined` in the return type. Both of these options are much better because it means the types will be correct. This means that that not only will the condition checks for our rules be right - it also will mean TS can catch many other classes of bugs automatically.

**Comment by atombrenner** [comment]
@bradzacher Totally agree and understand your arguments. The problem is, if you have an existing code base, and enable the strict ruleset, you get a lot of false warnings and many people just follow the eslint advice to remove the check.
Do you think we could improve the documentation of the rule? Example
- works only correct with direct index access or `noUncheckedIndexedAccess` or `arr.at(idx)`
- false positive will be reported when using destructuring or variable aliasing

Happy to contribute a PR :-)

**Comment by bradzacher** [comment]
Whilst the array case is not explicitly stated, it is broadly covered by the first paragraph in this section in the rule's docs

https://typescript-eslint.io/rules/no-unnecessary-condition/#when-not-to-use-it

The array case is a case of incorrect types (the undefined is not present in the types). And thus using a disable comment will make sense.

Do you think there is value in explicitly calling out the array case?

**Comment by kirkwaiblinger** [comment]
I could see it being a useful example of a very common case where the types will be possibly-incorrect.

**Comment by asburnin** [comment]
agree with @atombrenner here
when the rule can be false positive in some cases and forces to remove the necessary checks, potentially causing runtime errors — it totally deserves at least a mention in the documentation
